File: go/mask.go
package scopemask

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"sync"

	sqids "github.com/sqids/sqids-go"
)

const MinLength = 16
const BaseAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

const version = 1
const macLen = 4

var ErrInvalidID = errors.New("scopemask: invalid id")

const (
	tagInt   = 0
	tagStr   = 1
	tagBytes = 2
	tagUUID  = 3
)

const chunk = 6

type Integer interface {
	int | int8 | int16 | int32 | int64 |
		uint | uint8 | uint16 | uint32 | uint64
}

type Maskable interface {
	Integer | string | []byte | UUID
}

type UUID [16]byte

func ParseUUID(s string) (UUID, error) {
	s = strings.ReplaceAll(s, "-", "")
	if len(s) != 32 {
		return UUID{}, fmt.Errorf("scopemask: invalid UUID %q", s)
	}
	b, err := hex.DecodeString(s)
	if err != nil {
		return UUID{}, fmt.Errorf("scopemask: invalid UUID %q: %w", s, err)
	}
	var u UUID
	copy(u[:], b)
	return u, nil
}

func (u UUID) String() string {
	return fmt.Sprintf("%x-%x-%x-%x-%x", u[0:4], u[4:6], u[6:8], u[8:10], u[10:16])
}

func toPayload(value any) ([]byte, error) {
	var u uint64
	switch v := value.(type) {
	case int:
		if v < 0 {
			return nil, fmt.Errorf("int must be non-negative, got %d", v)
		}
		u = uint64(v)
	case int8:
		if v < 0 {
			return nil, fmt.Errorf("int must be non-negative, got %d", v)
		}
		u = uint64(v)
	case int16:
		if v < 0 {
			return nil, fmt.Errorf("int must be non-negative, got %d", v)
		}
		u = uint64(v)
	case int32:
		if v < 0 {
			return nil, fmt.Errorf("int must be non-negative, got %d", v)
		}
		u = uint64(v)
	case int64:
		if v < 0 {
			return nil, fmt.Errorf("int must be non-negative, got %d", v)
		}
		u = uint64(v)
	case uint:
		u = uint64(v)
	case uint8:
		u = uint64(v)
	case uint16:
		u = uint64(v)
	case uint32:
		u = uint64(v)
	case uint64:
		u = v
	case string:
		return append([]byte{tagStr}, v...), nil
	case []byte:
		return append([]byte{tagBytes}, v...), nil
	case UUID:
		return append([]byte{tagUUID}, v[:]...), nil
	default:
		return nil, fmt.Errorf("unsupported type for Encode: %T", value)
	}
	return append([]byte{tagInt}, minimalBE(u)...), nil
}

func fromPayload(p []byte) (any, error) {
	if len(p) == 0 {
		return nil, fmt.Errorf("%w: empty payload", ErrInvalidID)
	}
	tag, body := p[0], p[1:]
	switch tag {
	case tagInt:
		var n uint64
		for _, b := range body {
			n = n<<8 | uint64(b)
		}
		return n, nil
	case tagStr:
		return string(body), nil
	case tagBytes:
		return body, nil
	case tagUUID:
		if len(body) != 16 {
			return nil, fmt.Errorf("%w: bad UUID length %d", ErrInvalidID, len(body))
		}
		var u UUID
		copy(u[:], body)
		return u, nil
	default:
		return nil, fmt.Errorf("%w: unknown type tag %d", ErrInvalidID, tag)
	}
}

func minimalBE(v uint64) []byte {
	var b []byte
	for v > 0 {
		b = append([]byte{byte(v & 0xff)}, b...)
		v >>= 8
	}
	return b
}

func wrapPayload(secret []byte, scope string, inner []byte) []byte {
	signed := append([]byte{version}, inner...)
	return append(signed, computeChecksum(secret, scope, signed)...)
}

func computeChecksum(secret []byte, scope string, signed []byte) []byte {
	h := hmac.New(sha256.New, secret)
	h.Write([]byte(scope))
	h.Write([]byte{0})
	h.Write(signed)
	return h.Sum(nil)[:macLen]
}

func payloadToNums(p []byte) []uint64 {
	nums := []uint64{uint64(len(p))}
	for i := 0; i < len(p); i += chunk {
		end := i + chunk
		if end > len(p) {
			end = len(p)
		}
		var n uint64
		for _, b := range p[i:end] {
			n = n<<8 | uint64(b)
		}
		nums = append(nums, n)
	}
	return nums
}

func numsToPayload(nums []uint64) ([]byte, error) {
	if len(nums) == 0 {
		return nil, fmt.Errorf("%w: empty payload", ErrInvalidID)
	}
	length := int(nums[0])
	out := make([]byte, 0, length)
	remaining := length
	for _, n := range nums[1:] {
		k := chunk
		if remaining < k {
			k = remaining
		}
		if k < 8 && n>>(8*k) != 0 {
			return nil, fmt.Errorf("%w: chunk overflow", ErrInvalidID)
		}
		for i := k - 1; i >= 0; i-- {
			out = append(out, byte(n>>(8*i)))
		}
		remaining -= k
	}
	if remaining != 0 || len(out) != length {
		return nil, fmt.Errorf("%w: length mismatch", ErrInvalidID)
	}
	return out, nil
}

func castTo[T Maskable](zero T, raw any, scope, value string) (T, error) {
	mismatch := func() (T, error) {
		return zero, fmt.Errorf("%w for scope %q: %q is not the requested type", ErrInvalidID, scope, value)
	}
	switch any(zero).(type) {
	case string:
		s, ok := raw.(string)
		if !ok {
			return mismatch()
		}
		return any(s).(T), nil
	case []byte:
		b, ok := raw.([]byte)
		if !ok {
			return mismatch()
		}
		return any(b).(T), nil
	case UUID:
		u, ok := raw.(UUID)
		if !ok {
			return mismatch()
		}
		return any(u).(T), nil
	}
	n, ok := raw.(uint64)
	if !ok {
		return mismatch()
	}
	var out any
	switch any(zero).(type) {
	case int:
		out = int(n)
	case int8:
		out = int8(n)
	case int16:
		out = int16(n)
	case int32:
		out = int32(n)
	case int64:
		out = int64(n)
	case uint:
		out = uint(n)
	case uint8:
		out = uint8(n)
	case uint16:
		out = uint16(n)
	case uint32:
		out = uint32(n)
	case uint64:
		out = n
	default:
		return mismatch()
	}
	return out.(T), nil
}

func DeriveAlphabet(secret []byte, scope, base string) string {
	chars := []rune(base)
	var pool []byte
	pos, counter := 0, 0
	for i := len(chars) - 1; i > 0; i-- {
		if pos >= len(pool) {
			mac := hmac.New(sha256.New, secret)
			mac.Write([]byte(fmt.Sprintf("%s:%d", scope, counter)))
			pool = mac.Sum(nil)
			counter++
			pos = 0
		}
		j := int(pool[pos]) % (i + 1)
		pos++
		chars[i], chars[j] = chars[j], chars[i]
	}
	return string(chars)
}

type Option func(*ScopeMask)

func WithMinLength(n uint8) Option { return func(m *ScopeMask) { m.minLength = n } }
func WithBaseAlphabet(a string) Option {
	return func(m *ScopeMask) { m.baseAlphabet = a }
}

func WithPreviousSecrets(secrets ...string) Option {
	return func(m *ScopeMask) {
		for _, s := range secrets {
			if s != "" {
				m.secrets = append(m.secrets, []byte(s))
			}
		}
	}
}

// ScopeMask provides a way to mask integers, strings, and UUIDs
//
// It uses a combination of sqids and hmac to generate opaque ids that are
// scope-bound and secret-bound.
type ScopeMask struct {
	secrets      [][]byte
	minLength    uint8
	baseAlphabet string

	mu    sync.Mutex
	cache map[string]*sqids.Sqids
}

func New(secret string, opts ...Option) (*ScopeMask, error) {
	if secret == "" {
		return nil, errors.New("scopemask: secret must be non-empty")
	}
	m := &ScopeMask{
		secrets:      [][]byte{[]byte(secret)},
		minLength:    MinLength,
		baseAlphabet: BaseAlphabet,
		cache:        make(map[string]*sqids.Sqids),
	}
	for _, opt := range opts {
		opt(m)
	}
	return m, nil
}

func (m *ScopeMask) AlphabetFor(scope string) string {
	return DeriveAlphabet(m.secrets[0], scope, m.baseAlphabet)
}

func (m *ScopeMask) sqidsForIdx(idx int, scope string) (*sqids.Sqids, error) {
	key := fmt.Sprintf("%d\x00%s", idx, scope)
	m.mu.Lock()
	defer m.mu.Unlock()
	if s, ok := m.cache[key]; ok {
		return s, nil
	}
	s, err := sqids.New(sqids.Options{
		Alphabet:  DeriveAlphabet(m.secrets[idx], scope, m.baseAlphabet),
		MinLength: m.minLength,
		Blocklist: []string{},
	})
	if err != nil {
		return nil, err
	}
	m.cache[key] = s
	return s, nil
}

// Encode encodes a value to masked string.
func Encode[T Maskable](m *ScopeMask, scope string, value T, prefix string) (string, error) {
	inner, err := toPayload(any(value))
	if err != nil {
		return "", err
	}
	s, err := m.sqidsForIdx(0, scope)
	if err != nil {
		return "", err
	}
	id, err := s.Encode(payloadToNums(wrapPayload(m.secrets[0], scope, inner)))
	if err != nil {
		return "", err
	}
	return prefix + id, nil
}

// EncodeMany encodes a list of values to masked strings.
func EncodeMany[T Maskable](m *ScopeMask, scope string, values []T, prefix string) ([]string, error) {
	out := make([]string, len(values))
	for i, v := range values {
		id, err := Encode(m, scope, v, prefix)
		if err != nil {
			return nil, fmt.Errorf("index %d: %w", i, err)
		}
		out[i] = id
	}
	return out, nil
}

// Decode decodes a masked string to its original value.
func Decode[T Maskable](m *ScopeMask, scope, value, prefix string) (T, error) {
	return decode[T](m, scope, value, prefix)
}

// TryDecode attempts to decode a masked string to its original value.
func TryDecode[T Maskable](m *ScopeMask, scope, value, prefix string) (val T, ok bool) {
	v, err := decode[T](m, scope, value, prefix)
	return v, err == nil
}

// DecodeMany decodes a list of masked strings to their original values.
func DecodeMany[T Maskable](m *ScopeMask, scope string, values []string, prefix string) ([]T, error) {
	out := make([]T, len(values))
	for i, value := range values {
		v, err := Decode[T](m, scope, value, prefix)
		if err != nil {
			return nil, fmt.Errorf("index %d: %w", i, err)
		}
		out[i] = v
	}
	return out, nil
}

// TryDecodeMany attempts to decode a list of masked strings to their original values.
func TryDecodeMany[T Maskable](m *ScopeMask, scope string, values []string, prefix string) ([]T, []bool) {
	out := make([]T, len(values))
	oks := make([]bool, len(values))
	for i, value := range values {
		out[i], oks[i] = TryDecode[T](m, scope, value, prefix)
	}
	return out, oks
}

// Scope is a ScopeMask handle bound to one scope and prefix. Pass it to the
// *In functions instead of repeating the scope and prefix on every call.
type Scope struct {
	mask   *ScopeMask
	scope  string
	prefix string
}

// Scope returns a handle bound to one scope and prefix.
func (m *ScopeMask) Scope(scope, prefix string) Scope {
	return Scope{mask: m, scope: scope, prefix: prefix}
}

// EncodeIn encodes a value to masked string within a bound scope.
func EncodeIn[T Maskable](s Scope, value T) (string, error) {
	return Encode(s.mask, s.scope, value, s.prefix)
}

// EncodeManyIn encodes a list of values to masked strings within a bound scope.
func EncodeManyIn[T Maskable](s Scope, values []T) ([]string, error) {
	return EncodeMany(s.mask, s.scope, values, s.prefix)
}

// DecodeIn decodes a masked string to its original value within a bound scope.
func DecodeIn[T Maskable](s Scope, value string) (T, error) {
	return Decode[T](s.mask, s.scope, value, s.prefix)
}

// DecodeManyIn decodes a list of masked strings to their original values within a bound scope.
func DecodeManyIn[T Maskable](s Scope, values []string) ([]T, error) {
	return DecodeMany[T](s.mask, s.scope, values, s.prefix)
}

// TryDecodeIn attempts to decode a masked string to its original value within a bound scope.
func TryDecodeIn[T Maskable](s Scope, value string) (T, bool) {
	return TryDecode[T](s.mask, s.scope, value, s.prefix)
}

// TryDecodeManyIn attempts to decode a list of masked strings to their original values within a bound scope.
func TryDecodeManyIn[T Maskable](s Scope, values []string) ([]T, []bool) {
	return TryDecodeMany[T](s.mask, s.scope, values, s.prefix)
}

func decode[T Maskable](m *ScopeMask, scope, value, prefix string) (T, error) {
	var zero T
	if value == "" {
		return zero, fmt.Errorf("%w: empty id", ErrInvalidID)
	}
	if prefix != "" {
		if len(value) < len(prefix) || value[:len(prefix)] != prefix {
			return zero, fmt.Errorf("%w for scope %q: expected prefix %q", ErrInvalidID, scope, prefix)
		}
		value = value[len(prefix):]
	}
	for i, secret := range m.secrets {
		s, err := m.sqidsForIdx(i, scope)
		if err != nil {
			return zero, err
		}
		nums := s.Decode(value)
		if len(nums) == 0 {
			continue
		}
		reencoded, err := s.Encode(nums)
		if err != nil || reencoded != value {
			continue
		}
		payload, err := numsToPayload(nums)
		if err != nil || len(payload) < 2+macLen {
			continue
		}
		signed, tag := payload[:len(payload)-macLen], payload[len(payload)-macLen:]
		if signed[0] != version || !hmac.Equal(tag, computeChecksum(secret, scope, signed)) {
			continue
		}
		raw, err := fromPayload(signed[1:])
		if err != nil {
			continue
		}
		return castTo[T](zero, raw, scope, value)
	}
	return zero, fmt.Errorf("%w for scope %q", ErrInvalidID, scope)
}

File: python/scopemask/mask.py
import hashlib
import hmac
import uuid

from sqids import Sqids


MIN_LENGTH = 16
BASE_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

VERSION = 1
MAC_LEN = 4

TAG_INT = 0
TAG_STR = 1
TAG_BYTES = 2
TAG_UUID = 3

CHUNK = 6

Maskable = int | str | bytes | uuid.UUID


class InvalidId(Exception):
    pass


def to_payload(value: Maskable) -> bytes:
    if isinstance(value, bool):
        raise ValueError("bool is not supported; use int explicitly")
    if isinstance(value, int):
        if value < 0:
            raise ValueError(f"int must be non-negative, got {value!r}")
        nbytes = (value.bit_length() + 7) // 8
        return bytes([TAG_INT]) + value.to_bytes(nbytes, "big")
    if isinstance(value, uuid.UUID):
        return bytes([TAG_UUID]) + value.bytes
    if isinstance(value, str):
        return bytes([TAG_STR]) + value.encode("utf-8")
    if isinstance(value, (bytes, bytearray)):
        return bytes([TAG_BYTES]) + bytes(value)
    raise ValueError(f"unsupported type for encode: {type(value).__name__}")


def from_payload(inner: bytes) -> Maskable:
    if not inner:
        raise InvalidId("empty payload")
    tag, body = inner[0], inner[1:]
    if tag == TAG_INT:
        return int.from_bytes(body, "big")
    if tag == TAG_STR:
        return body.decode("utf-8")
    if tag == TAG_BYTES:
        return body
    if tag == TAG_UUID:
        if len(body) != 16:
            raise InvalidId(f"bad UUID length {len(body)}")
        return uuid.UUID(bytes=body)
    raise InvalidId(f"unknown type tag: {tag}")


def compute_checksum(secret: bytes, scope: str, signed: bytes) -> bytes:
    h = hmac.new(secret, scope.encode("utf-8") + b"\x00" + signed, hashlib.sha256)
    return h.digest()[:MAC_LEN]


def wrap_payload(secret: bytes, scope: str, inner: bytes) -> bytes:
    signed = bytes([VERSION]) + inner
    return signed + compute_checksum(secret, scope, signed)


def payload_to_nums(payload: bytes) -> list[int]:
    nums = [len(payload)]
    for i in range(0, len(payload), CHUNK):
        nums.append(int.from_bytes(payload[i : i + CHUNK], "big"))
    return nums


def nums_to_payload(nums: list[int]) -> bytes:
    if not nums:
        raise InvalidId("empty payload")
    length, out, remaining = nums[0], bytearray(), nums[0]
    for n in nums[1:]:
        k = min(CHUNK, remaining)
        try:
            out += n.to_bytes(k, "big")
        except OverflowError as exc:
            raise InvalidId("chunk overflow") from exc
        remaining -= k
    if remaining != 0 or len(out) != length:
        raise InvalidId("length mismatch")
    return bytes(out)


def as_bytes(secret: str | bytes) -> bytes:
    return secret.encode() if isinstance(secret, str) else secret


class ScopeMask:
    """
    ScopeMask provides a way to mask integers, strings, and UUIDs

    It uses a combination of sqids and hmac to generate opaque ids that are
    scope-bound and secret-bound.

    Args:
        secret (str | bytes): Secret key for masking.
        previous_secrets (tuple[str | bytes, ...], optional): Previous secrets for
            backward compatibility.
        min_length (int, optional): Minimum length of the masked id.
        base_alphabet (str, optional): Base alphabet for sqids.
    """
    def __init__(
        self,
        secret: str | bytes,
        *,
        previous_secrets: tuple[str | bytes, ...] = (),
        min_length: int = MIN_LENGTH,
        base_alphabet: str = BASE_ALPHABET,
    ):
        if not secret:
            raise ValueError("ScopeMask secret must be a non-empty string or bytes")
        self._secrets = [as_bytes(secret)] + [
            as_bytes(s) for s in previous_secrets if s
        ]
        self._min_length = min_length
        self._base_alphabet = base_alphabet
        self._sqids_cache: dict[tuple[int, str], Sqids] = {}

    @property
    def secret(self) -> bytes:
        return self._secrets[0]

    @property
    def base_alphabet(self) -> str:
        return self._base_alphabet

    def alphabet_for(self, scope: str, idx: int = 0) -> str:
        secret = self._secrets[idx]
        chars = list(self._base_alphabet)
        pool, pos, counter = b"", 0, 0
        for i in range(len(chars) - 1, 0, -1):
            if pos >= len(pool):
                pool = hmac.new(
                    secret, f"{scope}:{counter}".encode(), hashlib.sha256
                ).digest()
                counter += 1
                pos = 0
            j = pool[pos] % (i + 1)
            pos += 1
            chars[i], chars[j] = chars[j], chars[i]
        return "".join(chars)

    def sqids_for(self, idx: int, scope: str) -> Sqids:
        key = (idx, scope)
        sqids = self._sqids_cache.get(key)
        if sqids is None:
            sqids = Sqids(
                alphabet=self.alphabet_for(scope, idx),
                min_length=self._min_length,
                blocklist=[],
            )
            self._sqids_cache[key] = sqids
        return sqids


    def encode(
        self, scope: str, value: Maskable | None, prefix: str = ""
    ) -> str | None:
        """
        Encodes a value to masked string.

        Args:
            scope (str): Scope of the value.
            value (Maskable | None): Value to encode.
            prefix (str, optional): Prefix to add to the masked id.
        
        Returns:
            str | None: Masked id.
        """
        if value is None:
            return None
        payload = wrap_payload(self._secrets[0], scope, to_payload(value))
        nums = payload_to_nums(payload)
        return f"{prefix}{self.sqids_for(0, scope).encode(nums)}"

    def encode_many(
        self, scope: str, values: list[Maskable | None], prefix: str = ""
    ) -> list[str | None]:
        """
        Encodes a list of values to masked strings.

        Args:
            scope (str): Scope of the values.
            values (list[Maskable | None]): List of values to encode.
            prefix (str, optional): Prefix to add to the masked ids.
        
        Returns:
            list[str | None]: List of masked ids.
        """
        return list(map(lambda value: self.encode(scope, value, prefix), values))


    def decode(self, scope: str, value: str, prefix: str = "") -> Maskable:
        """
        Decodes a masked string to its original value.

        Args:
            scope (str): Scope of the masked id.
            value (str): Masked id to decode.
            prefix (str, optional): Prefix to remove from the masked id.
        
        Returns:
            Maskable: Decoded value.

        Raises:
            InvalidId: If the masked id is invalid.
        """
        if not isinstance(value, str) or not value:
            raise InvalidId(f"Empty or non-string id: {value!r}")
        if prefix and not value.startswith(prefix):
            raise InvalidId(
                f"Invalid id for scope {scope!r}: expected prefix {prefix!r}, "
                f"got {value!r}"
            )
        body = value[len(prefix):] if prefix else value
        for idx, secret in enumerate(self._secrets):
            sqids = self.sqids_for(idx, scope)
            decoded = sqids.decode(body)
            if not decoded:
                continue
            try:
                if sqids.encode(decoded) != body:
                    continue
                payload = nums_to_payload(decoded)
            except (ValueError, InvalidId):
                continue
            if len(payload) < 2 + MAC_LEN:
                continue
            signed, tag = payload[:-MAC_LEN], payload[-MAC_LEN:]
            if signed[0] != VERSION:
                continue
            if not hmac.compare_digest(tag, compute_checksum(secret, scope, signed)):
                continue
            return from_payload(signed[1:])
        raise InvalidId(f"Invalid id for scope {scope!r}: {value!r}")


    def try_decode(
        self, scope: str, value: str, prefix: str = ""
    ) -> Maskable | None:
        """
        Attempts to decode a masked string to its original value.

        Args:
            scope (str): Scope of the masked id.
            value (str): Masked id to decode.
            prefix (str, optional): Prefix to remove from the masked id.
        
        Returns:
            Maskable | None: Decoded value if successful, None otherwise.
        """
        try:
            return self.decode(scope, value, prefix)
        except InvalidId:
            return None

    def decode_many(
        self, scope: str, values: list[str], prefix: str = ""
    ) -> list[Maskable]:
        """
        Decodes a list of masked strings to their original values.

        Args:
            scope (str): Scope of the masked ids.
            values (list[str]): List of masked ids to decode.
            prefix (str, optional): Prefix to remove from the masked ids.
        
        Returns:
            list[Maskable]: List of decoded values.
        """
        return list(map(lambda value: self.decode(scope, value, prefix), values))

    def try_decode_many(
        self, scope: str, values: list[str], prefix: str = ""
    ) -> list[Maskable | None]:
        """
        Attempts to decode a list of masked strings to their original values.

        Args:
            scope (str): Scope of the masked ids.
            values (list[str]): List of masked ids to decode.
            prefix (str, optional): Prefix to remove from the masked ids.

        Returns:
            list[Maskable | None]: List of decoded values, None for any invalid id.
        """
        return list(map(lambda value: self.try_decode(scope, value, prefix), values))

    def scope(self, scope: str, prefix: str = "") -> "Scope":
        """
        Returns a handle bound to one scope and prefix.

        Args:
            scope (str): Scope to bind.
            prefix (str, optional): Prefix to bind.

        Returns:
            Scope: A handle whose encode/decode methods use the bound scope and prefix.
        """
        return Scope(self, scope, prefix)


class Scope:
    """
    A ScopeMask handle bound to one scope and prefix.

    It exposes the same encode and decode methods as ScopeMask, but the scope
    and prefix are fixed when the handle is created instead of passed each call.
    """

    def __init__(self, mask: ScopeMask, scope: str, prefix: str = ""):
        self._mask = mask
        self._scope = scope
        self._prefix = prefix

    def encode(self, value: Maskable | None) -> str | None:
        """
        Encodes a value to masked string.

        Args:
            value (Maskable | None): Value to encode.

        Returns:
            str | None: Masked id.
        """
        return self._mask.encode(self._scope, value, self._prefix)

    def encode_many(self, values: list[Maskable | None]) -> list[str | None]:
        """
        Encodes a list of values to masked strings.

        Args:
            values (list[Maskable | None]): List of values to encode.

        Returns:
            list[str | None]: List of masked ids.
        """
        return self._mask.encode_many(self._scope, values, self._prefix)

    def decode(self, value: str) -> Maskable:
        """
        Decodes a masked string to its original value.

        Args:
            value (str): Masked id to decode.

        Returns:
            Maskable: Decoded value.

        Raises:
            InvalidId: If the masked id is invalid.
        """
        return self._mask.decode(self._scope, value, self._prefix)

    def decode_many(self, values: list[str]) -> list[Maskable]:
        """
        Decodes a list of masked strings to their original values.

        Args:
            values (list[str]): List of masked ids to decode.

        Returns:
            list[Maskable]: List of decoded values.
        """
        return self._mask.decode_many(self._scope, values, self._prefix)

    def try_decode(self, value: str) -> Maskable | None:
        """
        Attempts to decode a masked string to its original value.

        Args:
            value (str): Masked id to decode.

        Returns:
            Maskable | None: Decoded value if successful, None otherwise.
        """
        return self._mask.try_decode(self._scope, value, self._prefix)

    def try_decode_many(self, values: list[str]) -> list[Maskable | None]:
        """
        Attempts to decode a list of masked strings to their original values.

        Args:
            values (list[str]): List of masked ids to decode.

        Returns:
            list[Maskable | None]: List of decoded values, None for any invalid id.
        """
        return self._mask.try_decode_many(self._scope, values, self._prefix)
